x-pack/plugins/task_manager/server/lib/create_managed_configuration.test.ts

```diff
@@ -0,0 +1,213 @@
+    expect(pollIntervalSubscription).toHaveBeenCalledTimes(1);
+  });
+
+  describe('maxWorker configuration', () => {
+    function setupScenario(startingMaxWorkers: number) {
+      const errors$ = new Subject<Error>();
+      const subscription = jest.fn();
+      const { maxWorkersConfiguration$ } = createManagedConfiguration({
+        errors$,
+        startingMaxWorkers,
+        logger,
+        startingPollInterval: 1,
+      });
+      maxWorkersConfiguration$.subscribe(subscription);
+      return { subscription, errors$ };
+    }
+
+    beforeEach(() => {
+      jest.resetAllMocks();
+      clock = sinon.useFakeTimers();
+    });
+
+    afterEach(() => clock.restore());
+
+    test('should decrease configuration at the next interval when an error is emitted', async () => {
+      const { subscription, errors$ } = setupScenario(100);
+      errors$.next(SavedObjectsErrorHelpers.createTooManyRequestsError('a', 'b'));
+      clock.tick(ADJUST_THROUGHPUT_INTERVAL - 1);
+      expect(subscription).toHaveBeenCalledTimes(1);
+      clock.tick(1);
+      expect(subscription).toHaveBeenCalledTimes(2);
+      expect(subscription).toHaveBeenNthCalledWith(2, 80);
+    });
+
+    test('should log a warning when the configuration changes from the starting value', async () => {
+      const { errors$ } = setupScenario(100);
+      errors$.next(SavedObjectsErrorHelpers.createTooManyRequestsError('a', 'b'));
+      clock.tick(ADJUST_THROUGHPUT_INTERVAL);
+      expect(logger.warn).toHaveBeenCalledWith(
+        'Max workers configuration is temporarily reduced after Elasticsearch returned 1 "too many request" error(s).'
+      );
+    });
+
+    test('should increase configuration back to normal incrementally after an error is emitted', async () => {
+      const { subscription, errors$ } = setupScenario(100);
+      errors$.next(SavedObjectsErrorHelpers.createTooManyRequestsError('a', 'b'));
+      clock.tick(ADJUST_THROUGHPUT_INTERVAL * 10);
+      expect(subscription).toHaveBeenNthCalledWith(2, 80);
+      expect(subscription).toHaveBeenNthCalledWith(3, 84);
+      // 88.2- > 89 from Math.ceil
+      expect(subscription).toHaveBeenNthCalledWith(4, 89);
+      expect(subscription).toHaveBeenNthCalledWith(5, 94);
+      expect(subscription).toHaveBeenNthCalledWith(6, 99);
+      // 103.95 -> 100 from Math.min with starting value
+      expect(subscription).toHaveBeenNthCalledWith(7, 100);
+      // No new calls due to value not changing and usage of distinctUntilChanged()
+      expect(subscription).toHaveBeenCalledTimes(7);
+    });
+
+    test('should keep reducing configuration when errors keep emitting', async () => {
+      const { subscription, errors$ } = setupScenario(100);
+      for (let i = 0; i < 20; i++) {
+        errors$.next(SavedObjectsErrorHelpers.createTooManyRequestsError('a', 'b'));
+        clock.tick(ADJUST_THROUGHPUT_INTERVAL);
+      }
+      expect(subscription).toHaveBeenNthCalledWith(2, 80);
```

Comment:
nit: could put these numbers in an array and do the `expect()`'s in a loop, but may be harder to debug problems that way ...

Reply:
I've been thinking about this as well. The two up sides I saw this way 1) it provides a clear example of how the configuration gets reduced from 100 when errors keep emitting 2) it allowed to add comments to explain some of the inner usage of Math.floor and distinctUntilChanged() as the assertions happened. I could always cut a few assertions out.